src/View/Form/ContextFactory.php

```diff
@@ -0,0 +1,143 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\View\Form;
+
+use Cake\Collection\Collection;
+use Cake\Datasource\EntityInterface;
+use Cake\Form\Form;
+use Cake\Http\ServerRequest;
+use RuntimeException;
+use Traversable;
+
+/**
+ * Factory for getting form context instance based on provided data.
+ */
+class ContextFactory
+{
+    /**
+     * Context provider methods.
+     *
+     * @var array
+     */
+    protected $contextProviders = [];
+
+    /**
+     * Constructor.
+     *
+     * @param array $providers Array of provider callables. Each element should
+     *   be of form `['type' => 'a-string', 'callable' => ..]`
+     * @param bool $addDefaults Whether default providers should be added.
+     */
+    public function __construct(array $providers = [], $addDefaults = true)
```

Comment:
More that `ContextFactory` could have a method like `ContextFactory::createWithDefaults()`.